### ISSUE TITLE
Work without PulseAudio

### ISSUE THREAD

**Issue by ndarilek** [MycroftAI/mycroft-core#1676]
I'm attempting to build a Mycroft addon for Home Assistant's Hass.io container-based distribution. Unlike my experience doing this for previous RPi models, Hass.io's audio works for me out-of-the-box. Whereas before I needed PulseAudio to keep my HDMI audio open to not cut off the beginning of spoken utterances, Hass.io provides flawless spoken feedback without Pulse. It also makes installing PA very difficult since everything is done within containers, and I'd really like to avoid adding an audio server if it isn't needed.

Unfortunately, while Mycroft TTS works flawlessly with no additional configuration, I can't get the wake word to work at all. I've confirmed that arecord captures audio from the mic by default, with no additional configuration. For some reason, though, Mycroft seems to require PA for input but not for output.

Is there a configuration setting I might use to tell Mycroft to use Alsa's default capture device rather than trying to go through PA? Pulse will unfortunately be a non-starter for this configuration, and again, everything else (including Mycroft's TTS) is working fine without it. I understand that audio configuration is a tricky thing to get right, but Hass.io lets you configure the default audio inputs/outputs for a given addon from its dashboard. Seems like, if I can hand Mycroft a working set of defaults, that I should at least be able to configure it to use *those* and skip PA entirely.

Thanks.

**Comment by forslund** [comment]
As far as I know there's no hard requirement for pulseaudio, The input stack for mycroft is built on the shoulders of giants:
Mycroft->speech_recognition->PythonAudio->PortAudio

And PortAudio should be able to work directly on top of Alsa. That said, I've not actually tried it myself.

```python
import pyaudio
p = pyaudio.PyAudio()
for i in range(p.get_host_api_info_by_index(0).get('deviceCount')):
        if (p.get_device_info_by_host_api_device_index(0, i).get('maxInputChannels')) > 0:
            print("Input Device id ", i, " - ", p.get_device_info_by_host_api_device_index(0, i).get('name'))
```
Which should list the input devices found.

If that lists a device you could try to force the usage of that device by setting the `device_index` parameter in the listener part of the config (can for example be created as `/etc/mycroft/mycroft.conf`:

```json
{
  "listener": {
    "device_index": X
  }
}
```

Where X is the id reported by the script above.

This is all theory and I might be wrong.

**Comment by ndarilek** [comment]
Having a hard time getting this working with PulseAudio on my custom platform, and am considering just going direct. What would I lose if I did this?

It looks like Mycroft may duck non-voice audio if everything is using Pulse. Is that accurate or am I misremembering/misreading?

Thanks.

**Comment by forslund** [comment]
No pulseaudio features are used by default so you won't lose anything.

(The pulse audio ducking needs to be enabled in config.)

**Comment by ndarilek** [comment]
Yeah, that's how I got the 2 to which I set the device index.

I do still have a PULSE_SERVER environment variable set. I'll try 
removing that in case the library is trying to connect to my 
no-longer-running Pulse server, or is using that to determine how many 
devices are available. That just occurred to me, been trying so many 
things to get this working that I can't keep everything's state straight.

**Comment by forslund** [comment]
Right sorry, missed that line :)

Might also be an issue with speech recognition. I'll try to get a docker image going as well and see if I can help.

**Comment by ndarilek** [comment]
Removing PULSE_SERVER didn't change anything. If I remove the listener 
config, the error goes away as well, but input still doesn't work.

I've just confirmed that `arecord test.wav ... aplay test.wav` works 
(I.e. no need to specify recording or playback devices.) So the issue 
seems to be that Mycroft isn't detecting the default recording device 
for some reason.

My setup is a Docker container running the latest Debian armhf packages 
on an RPi 3 with a USB camera mic and the above /etc/asound.conf setting 
up the defaults. 
[Here](https://gitlab.com/ndarilek/hassio-mycroft/tree/master/mycroft) 
is how I'm setting it up to run in Hassio.

Thanks!

**Comment by forslund** [comment]
Hmm, I don't get that error. I get an invalid sample rate error instead....

**Comment by ndarilek** [comment]
OK, I appear to have cracked it. Was a combination of forgetting to give the Mycroft addon audio permissions (previously Mycroft used Pulse and the Pulse addon had audio permissions) and removing some out-of-date configuration in my local config.

Thanks for the help! I'll see if I can get this packaged and available for the community in the next few days.

**Comment by ndarilek** [comment]
Figured out how to make Pulse show up as an input device. I needed the libasound2-plugins package and the following ~/.asoundrc:

```
pcm.pulse {
    type pulse
}

ctl.pulse {
    type pulse
}
```

But I'm officially giving up on trying to make this work without Pulse. Looks like PyAudio has [many](https://stackoverflow.com/questions/43947294/i-have-trouble-with-pyaudio-does-not-detect-input-device) [issues](https://raspberrypi.stackexchange.com/questions/59852/pyaudio-does-not-detect-my-microphone-connected-via-usb-audio-adapter) finding default input devices, and I was able to containerize a PulseAudio install. With Pulse, the correct asoundrc configuration, and libasound2-plugins, I appear to have audio somewhat working, though I'm hoping to tune it further. Without it, pyaudio just doesn't detect my input device anymore, and I have no clue why this worked for a time but now fails.
